### PR TITLE
Add invoice notifications

### PR DESCRIPTION
**Pull request by Kukks**
pav needs to play around and see which ones fit and what needs to be edited

### CODE REVIEW

File: BTCPayServer/Services/Notifications/Blobs/InvoiceEventNotification.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                {InvoiceEvent.FailedToConfirm, "has payments that failed to confirm on time."},
+                {InvoiceEvent.ReceivedPayment, "received a payment."},
+                {InvoiceEvent.Confirmed, "was confirmed."}
+            };
+            
+            protected override void FillViewModel(InvoiceEventNotification notification,
+                NotificationViewModel vm)
+            {
+                var baseStr = $"Invoice {notification.InvoiceId.Substring(0, 5)}..";
+                vm.Body = $"{TextMapping[notification.Event]} {baseStr}";
```
Comment (rockstardev) [inline BTCPayServer/Services/Notifications/Blobs/InvoiceEventNotification.cs:40]: You've missed the ordering of vars, should be `$"{baseStr} {TextMapping[notification.Event]}"`. 

I would invite @pavlenex to double check and verify strings.

**Comment by Kukks** [comment]
> 
> 
> PR looks great. Design wise, we need to get exact look that would work for all notifications - taking into consideration that most of them will be textual and not necessary have icon involved. You @pavlenex can tell us if we can tap designers that already worked with project, or 3 of us will put our design hat on and sketch it.
> 
> On colors for certain events - I guess we should leverage styling already present in themes. If we need to go beyond that - again - it's design effort to specify how those colors map to notifications, and can then add that info to `TextMapping` variable.

I was thinking we could maybe add a `BodyClass` property for the VM. Then for specific things, we could try and use `alert-warning`, etc as our pav overlord deems fit

**Comment by rockstardev** [comment]
And I think this is good change: "Instead of date use Ago like we do for invoices (I think ago is more relevant for notifications)" 

@Kukks you can just switch from `ToBrowserDate()` to `ToTimeAgo()` in NotificationNavItem.cshtml#22... I would do it but don't have your fork as remote registered.

**Comment by rockstardev** [comment]
> I was thinking we could maybe add a BodyClass property for the VM. Then for specific things, we could try and use alert-warning, etc as our pav overlord deems fit

I think this is why "we need to get exact look that would work for all notifications". It would allow us to clear unknowns and know if we need `BodyClass` or we'll require even a bigger weapon (even more control over resulting HTML).

**Comment by pavlenex** [comment]
> @pavlenex can tell us if we can tap designers that already worked with project, or 3 of us will put our design hat on and sketch it.

I'm afraid we won't have the time to involve designers for the next release, looking at the progress we had in past on design, it always took few weeks. If you guys want, we can improve the coloring or icons or whatever in the next release, but for now, we can focus on creating a clean, simple design which we can expand in the future. However if you want to tackle design, I think this one is quite [solid](https://user-images.githubusercontent.com/36959754/85150574-fe326780-b252-11ea-9630-b2caf654edaa.PNG). 

One reason I want to have different coloring involved is because not many people especially when they begin with BTCPay know the difference between paid and completed. They see paid and assume it's okay, if we can tackle that one textually somehow, for now I'd be good with it.

**Comment by pavlenex** [comment]
Testing the PR after 5f2c5da it now looks solid. So the two concerns we should address before merging this:
- Adjusting notification appearance for themes.
- Distinguishing paid and complete

Capture

@rockstardev @Kukks how would you like to distinguish the `paid` and `completed` status? If we want to do it textually for now, I suggest we add something suggested in #1319
Perhaps:

`Invoice paid, pending confirmation`
`Invoice completed` or `Invoice paid and confirmed` or `Invoice fully paid and confirmed`

**Comment by NicolasDorier** [comment]
This is too spammy for users. Please for now, can you just put a notification when it get paid?

**Comment by Kukks** [comment]
> 
> 
> This is too spammy for users. Please for now, can you just put a notification when it get paid?

What about dropping `ReceivedPayment` and `PaidInFull` and leaving the rest

This is how it looks on a an invoice being paid and confirmed on time:
image


The rest only show in invalid situations where a merchant needs to take action

**Comment by pavlenex** [comment]
> 
> 
> This is too spammy for users. Please for now, can you just put a notification when it get paid?

What is spammy? Complete and Paid or also invalid and partially paid? I think we can go with `complete` but I would not drop `invalid` and `partiallypaid` it's important to be notified of those and they don't happen that often anyways.

**Comment by NicolasDorier** [comment]
Sorry, actually seems fine. But I would remove "confirmed" or "fully paid".

**Comment by Kukks** [comment]
> 
> 
> Sorry, actually seems fine. But I would remove "confirmed" or "fully paid".

I removed ReceivedPayment and PaidInFull

This way, you know that an invoice has been paid and confirmed correctly

**Comment by NicolasDorier** [comment]
@Kukks anyway they will know by checking the invoice. What they want to know via notification is "wow I got money in"

**Comment by pavlenex** [comment]
I agree that initial "wow I got money" is important, especially in retail, but for online businesses that use integration it won't matter. I guess once we add notification preferences it'll make more sense for certain type of users.

**Pull request merged**